ADR for private cert management - OHSH-511 (#270)
OHS-Hosting-Infrastructure/infrastructure: docs/adr/0015-private-cert-management.md

  - [Table of Contents](#table-of-contents)
  - [Context and Problem Statement](#context-and-problem-statement)
  - [Decision Drivers](#decision-drivers)
  - [Considered Options](#considered-options)
  - [Decision Outcome](#decision-outcome)
  - [Pros and Cons of the Options](#pros-and-cons-of-the-options)
    - [Continue manually managing private certs](#continue-manually-managing-private-certs)
    - [Use SSLMate for generating and managing private certs](#use-sslmate-for-generating-and-managing-private-certs)
    - [Use private cert management through AWS](#use-private-cert-management-through-aws)
  - [Links](#links)

<!-- mdformat-toc end -->

## Context and Problem Statement

While working to renew an expired certificate on the iRedMail server, it was highlighted that we don't currently have a good way to generate/manage private certificates. At the moment it seems the kind of management employed differs server by server.

## Decision Drivers

- We need a standardized way to be alerted when new private certs need to be generated.
- We need a standardized way for generating new certs when needed.

> 

Reviewer: I would match the drivers to the common factors in Good/Bad in the options. You mention cost, ease of use/automation/consistency

## Considered Options

- Continue manually managing private certs
- Use SSLMate for generating and managing private certs
- Use private cert management through AWS

## Decision Outcome

> Using SSLMate for generating and managing private certificates seems to be the best option. It provides automation, alerts, and a web interface. SSLMate has also been previously vetted for use on the MilMove project.

Reviewer 2: ```suggestion
Using SSLMate for generating and managing private certificates. It provides automation, alerts, and a web interface. SSLMate has also been previously vetted for use on the MilMove project.
```

## Pros and Cons of the Options

### Continue manually managing private certs

> - Good, because it seems to be the current way of operating. Which provides a certain kind of ease/consistency.

Reviewer 2: ```suggestion
- Good, because it is the current way of operating. Which provides a certain kind of ease/consistency.
```

- Bad, because it quickly becomes difficult to track when certs are set to expire.
- Bad, because it's difficult to track which certs each app is using and for what at any given time.

> - Bad, because maintaining our own CA, or even multiple CAs, isn't advisable.

Reviewer 2: Can we say a little more about why not? Here, or in the decision drivers, or both?

### Use SSLMate for generating and managing private certs

- Good, because it provides alerting for certificate expiration.
- Good, because it provides alerting if something goes wrong during the renewal process
- Good, because it can automatically renew and install new certificates on to servers.
- Good, because it can automate the entire process of private cert generation.

> - Good, because it provides a central location for tracking and managing every private cert we would need.

Reviewer 3: Oh ! I didn't know! It's not through AWS or do we import it to AWS?

Author: yeah, SSLMate is it's own tool operating outside of our aws infrastructure. and as far as i'm aware we would be using either its cli or web console to manage the certs that we generate through them.

> - Good, because the cost scales by year. Rather by month.

Reviewer 2: ```suggestion
- Good, because the cost scales by year, rather by month.
```

Reviewer 3: Since you include cost of AWS private cert, would you mind including either the price or that it's a small enouhg amount that we can reasonably not worry about it?

- Bad, because it's another tool to add to learn and implement into our process.

### Use private cert management through AWS

- Good, because we would be remaining in the AWS ecosystem of products.
- Good, because private cert management would live in the same place we already manage our public certs.

> - Bad, because starting at $400 a month for each private CA required to generate new private certs is fairly prohibitive.

Reviewer 3: All of these are applicable to this one too: 

```
- Good, because it provides alerting for certificate expiration.
- Good, because it provides alerting if something goes wrong during the renewal process
- Good, because it can automatically renew and install new certificates on to servers.
- Good, because it can automate the entire process of private cert generation.
```

  - [Links](#links)

<!-- mdformat-toc end -->

## Context and Problem Statement

While working to renew an expired certificate on the iRedMail server, it was highlighted that we don't currently have a good way to generate/manage private certificates. At the moment it seems the kind of management employed differs server by server.

## Decision Drivers

- We need a standardized way to be alerted when new private certs need to be generated.
- We need a standardized way for generating new certs when needed.
- Being our own CA, as is the current set up on at least one server, isn't advisable from a availability/security perspective. Such a practice could leave our apps/services open to man in the middle attacks and users could receive warnings when trying to access one of the apps/services all due to us not being considered an trusted CA.
- The solution has to be affordable and provide a reliable way to track/manage our private certificates.
- Preferably the solution will have a low learning curve.

## Considered Options

- Continue manually managing private certs
- Use SSLMate for generating and managing private certs
- Use private cert management through AWS

## Decision Outcome

> Use AWS to generate and manage a single root CA, from which we can generate the needed private certificates. Seeing as our current setup of using non-qualified hostnames limits our options among third-party options.

Reviewer 2: ```suggestion
Use AWS to generate and manage a single root CA, from which we can generate the needed private certificates. Our use of non-qualified domain names limits third-party options, as they do not support this.
```

- Good, because it can automatically renew and install new certificates on to servers.
- Good, because it can automate the entire process of private cert generation.
- Good, because it provides a central location for tracking and managing every private cert we would need.
- Good, because the cost scales by year, rather by month. Starting at $15.95 a year for a single hostname, and $149.95 a year for a wildcard hostname.
- Bad, because it's another tool to add to learn and implement into our process.
- Bad, because we can't use non-qualified hostnames to generate certs.

### Use private cert management through AWS

- Good, because it provides alerting for certificate expiration.
- Good, because it provides alerting if something goes wrong during the renewal process
- Good, because it can automatically renew and install new certificates on to servers.
- Good, because it can automate the entire process of private cert generation.
- Good, because we would be remaining in the AWS ecosystem of products.
- Good, because private cert management would live in the same place we already manage our public certs.
- Good, because we can continue to use our current set of internal hostnames.

> - Bad, because starting at $400 a month for each private CA required to generate new private certs is fairly prohibitive.

Reviewer 2: Is it prohibitive? That seems like a trivial amount of money.

Author: hmmm let me rework this one
